### PR TITLE
Architecture n-tier 

### PR DESCRIPTION
**Pull request by vlad12345123**
Architecture n-tier implemented and added work with files

### CODE REVIEW

File: InternshipBE/ExaLearn.Bl/DTO/FileEntryDTO.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Bl.DTO
+{
+    public class FileEntryDTO
+    {
+        //public int Id { get; set; }
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/DTO/FileEntryDTO.cs]: remove comments and leave the fields

File: InternshipBE/ExaLearn.Bl/Extensions/AudioExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Bl.FileFormats
+{
+    internal static class AudioExtensions
+    {
+        public static readonly string[] Extensions =
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Extensions/AudioExtensions.cs]: rename

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        public FileService(IFileRepository filesRepository)
+        {
+            _filesRepository = filesRepository;
+        }
+
+        public async Task<FileEntry> AddAsync(IFormFile file)
+        {
+            /*
+             do not forget to clarify the point regarding the naming of files!
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: comment

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+             2.say that such a file exists
+             3.rewrite if names match
+
+            */
+
+            var fileType = Path.GetExtension(file.FileName);
+
+            var fileEntry = new FileEntry();
+
+            foreach (var format in AudioExtensions.Extensions)
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: Contains Linq

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            {
+                if (fileType.ToLower() == format) // realize this in enum
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: StringComparer

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                {
+                    var filePath = "/Files/Audio/" + file.FileName; //absolute path to audiofile
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: const

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+                    if (file != null && file.Length >= 0)
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: upper

Comment (RomanMRC) [reply InternshipBE/ExaLearn.Bl/Services/FileService.cs:44]: String.IsNullOrEmpty

File: InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿using ExaLearn.Dal.Model;
+using Microsoft.EntityFrameworkCore;
+
+namespace ExaLearn.Dal.Database
+{
+    public class ExaLearnDbContext : DbContext
+    {
+        public DbSet<FileEntry> FileEntry { get; set; }
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs]: bottom

File: InternshipBE/ExaLearn.Dal/DbInitializer/DbInitialize.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.DbInitializer
+{
+    public class DbInitialize 
+    {
+        public static void Initialize(ExaLearnDbContext exaLearnDbContext)
+        {
+            exaLearnDbContext.Database.EnsureCreated(); // its create db before deploy our project on the server
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/DbInitializer/DbInitialize.cs]: remove

File: InternshipBE/ExaLearn.Dal/Interfaces/IFileRepository.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Repositories
+{
+    public interface IFileRepository
+    {
+        Task<List<FileEntry>> GetAllFilesAsync();
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Interfaces/IFileRepository.cs]: spaces

File: InternshipBE/ExaLearn.Dal/Model/FileEntry.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Model
+{
+    public class FileEntry
+    {
+        public int Id { get; set; }
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Model/FileEntry.cs]: spaces

File: InternshipBE/ExaLearn.Dal/Repositories/FileRepository.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Repositories
+{
+    public class FileRepository : IFileRepository
+    {
+        private ExaLearnDbContext appDbContext;
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/FileRepository.cs]: space

File: InternshipBE/ExaLearn.Dal/Repositories/FileRepository.cs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        public FileRepository(ExaLearnDbContext appDbContext)
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/FileRepository.cs]: generic repository

File: InternshipBE/ExaLearn.Dal/Repositories/Repository.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Repositories
+{
+    class Repository
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/Repository.cs]: implement generic repository

File: InternshipBE/WebApi/Startup.cs [hunk 15]
```diff
@@ -1,16 +1,14 @@
...
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Hosting;
-using Microsoft.Extensions.Logging;
 using Microsoft.OpenApi.Models;
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
 
 namespace WebApi
```
Comment (romanstat) [inline InternshipBE/WebApi/Startup.cs]: ExaLearn.WebApi

File: InternshipBE/WebApi/Controllers/FilesController.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+        [HttpPost]
+        public async Task<IActionResult> CreateFileAsync(IFormFile file)
+        {
+            return Ok(await _fileService.AddAsync(file));
+        }
+
+
+        [HttpGet("{id}")]
+        public async Task<IActionResult> Get(int id)
```
Comment (ExadelSandbox) [inline InternshipBE/WebApi/Controllers/FilesController.cs]: Please, Avoid using the meaningless "Get" Does it mean Get files for listening or get file for estimation?

File: InternshipBE/WebApi/Controllers/FilesController.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace WebApi.Controllers
+{
+    [Route("[controller]")]
```
Comment (ExadelSandbox) [inline InternshipBE/WebApi/Controllers/FilesController.cs]: it had better to remove the Route attribute from here and provide routing for each methods

**Comment by vlad12345123** [comment]
I do changes!

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.Extensions.Configuration;
+using System.Linq;
+using ExaLearn.Dal.Interfaces;
+
+namespace ExaLearn.Bl.Services
+{
+    public class FileService : IFileService
+    {
+        IGenericRepository<FileEntry> _filesRepository;
+        IConfiguration _configuration; //it should be in startup.. but for now i do it like this
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: space between ctor and field

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        public async Task<FileEntry> AddAsync(IFormFile file)
+        {
+            var fileEntry = new FileEntry();
+
+            if (file != null && file.Length >= 0) 
+            {
+                var fileType = Path.GetExtension(file.FileName);
+                var isNessesaryFormat = AudioExtensions.AvalableAudutionExtensions.Any(e => e.Contains(fileType.ToLower()));
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: remove Any

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+                if (isNessesaryFormat)
+                {
+                    var filePath = _configuration.GetConnectionString("HostingContext");
+
+                    using (var stream = new FileStream(Path.Combine(filePath, Path.GetRandomFileName()), FileMode.Create)) //random filename or not?
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: remove comment

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+        public async Task<FileEntry> AddAsync(IFormFile file)
+        {
+            var fileEntry = new FileEntry();
+
+            if (file != null && file.Length >= 0) 
+            {
+                var fileType = Path.GetExtension(file.FileName);
+                var isNessesaryFormat = AudioExtensions.AvalableAudutionExtensions.Any(e => e.Contains(fileType.ToLower()));
+
+                if (isNessesaryFormat)
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: ! and throw ValidationException

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+                        await file.CopyToAsync(stream);
+                    }
+                    fileEntry.Name = file.FileName;
+                    fileEntry.Url = filePath;
+                }
+            }
+
+            return await _filesRepository.AddAsync(fileEntry);
+        }
+
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs:53]: remove space

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+
+        public async Task<FileEntry> GetAsync(FileEntry fileEntry)
+        {
+            return await _filesRepository.GetAsync(fileEntry);
+        }
+        public async Task<List<FileEntry>> GetFilesAsync()
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs:59]: add space

File: InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace ExaLearn.Dal.Database
+{
+    public class ExaLearnDbContext : DbContext
+    {
+        public ExaLearnDbContext(DbContextOptions<ExaLearnDbContext> options) : base(options)
+        {
+
+        }
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs:11]: space

File: InternshipBE/ExaLearn.Dal/Interfaces/IGenericRepository.cs [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Interfaces
+{
+    public interface IGenericRepository<T> where T : class
+    {
+
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Interfaces/IGenericRepository.cs]: refactoring

File: InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Repositories
+{
+    public class GenericRepository<T> : IGenericRepository<T> where T : class
+    {
+
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs:16]: remove space

File: InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+        private ExaLearnDbContext appDbContext;
+        private DbSet<T> dbSet;
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs]: add space

File: InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+        public GenericRepository(ExaLearnDbContext appDbContext)
+        {
+            this.appDbContext = appDbContext;
+            dbSet = appDbContext.Set<T>();
+        }
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs:21]: add space

File: InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Repositories
+{
+    public class GenericRepository<T> : IGenericRepository<T> where T : class
+    {
+
+        private ExaLearnDbContext appDbContext;
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs]: _appDbContext

File: InternshipBE/ExaLearn.Tests/UnitTest1.cs [hunk 30]
```diff
@@ -0,0 +1,14 @@
+using System;
+using Xunit;
+
+namespace ExaLearn.Tests
+{
+    public class UnitTest1
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Tests/UnitTest1.cs]: remove file

File: InternshipBE/WebApi/Controllers/FilesController.cs [hunk 31]
```diff
@@ -0,0 +1,35 @@
+﻿using ExaLearn.Bl.Interfaces;
+using ExaLearn.Dal.Model;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using System.Threading.Tasks;
+
+namespace WebApi.Controllers
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/FilesController.cs]: ExaLearn.WebApi

File: InternshipBE/WebApi/Controllers/FilesController.cs [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+{
+    [Route("[controller]")]
+    [ApiController]
+    public class FilesController : ControllerBase
+    {
+
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/FilesController.cs]: remove space

File: InternshipBE/WebApi/Controllers/FilesController.cs [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+            _fileService = fileService;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> CreateFileAsync(IFormFile file)
+        {
+            return Ok(await _fileService.AddAsync(file));
+        }
+
+
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/FilesController.cs]: remove space

File: InternshipBE/WebApi/Controllers/FilesController.cs [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+        [HttpGet("{id}")]
+        public async Task<IActionResult> Get(FileEntry fileEntry)
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/FilesController.cs]: FileEntry -> id
Get -> GetById

File: InternshipBE/ExaLearn.Bl/Extensions/AudioExtensions.cs [hunk 35]
```diff
@@ -0,0 +1,11 @@
+﻿
+namespace ExaLearn.Bl.FileFormats
+{
+    internal static class AudioExtensions
+    {
+        public static readonly string[] AvalableAudutionExtensions =
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Extensions/AudioExtensions.cs]: Available

File: InternshipBE/ExaLearn.Bl/Services/FileService.cs [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+using Microsoft.Extensions.Configuration;
+using System.Linq;
+using ExaLearn.Dal.Interfaces;
+using System.ComponentModel.DataAnnotations;
+
+namespace ExaLearn.Bl.Services
+{
+    public class FileService : IFileService
+    {
+        IGenericRepository<FileEntry> _filesRepository;
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Bl/Services/FileService.cs]: private readonly

File: InternshipBE/WebApi/Startup.cs [hunk 37]
```diff
@@ -32,9 +32,12 @@ public void ConfigureServices(IServiceCollection services)
             {
                 c.SwaggerDoc("v1", new OpenApiInfo { Title = "WebApi", Version = "v1" });
             });
+            //services.AddScoped(typeof(IGenericRepository<>), typeof(GenericRepository<>));
```
Comment (romanstat) [inline InternshipBE/WebApi/Startup.cs]: remove comment

**Review by romanstat — APPROVED** [review]

**Review by artemgael96 — APPROVED** [review]

File: InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs [hunk 38]
```diff
@@ -0,0 +1,53 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.Dal.Repositories
+{
+    public class GenericRepository<T> : IGenericRepository<T> where T : class
+    {
+        private ExaLearnDbContext _appDbContext;
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Repositories/GenericRepository.cs]: protected readonly

**Review by ExadelSandbox — APPROVED** [review]

**Pull request merged**